README.org
#+TITLE: Read, Process and Create iCalendar Data and Files
#+AUTHOR: Enrico Schumann

Utilities for handling icalendar data. The package is
written in pure R, with no strict dependencies (package
'base64enc' is only needed if calendar files have
attachments). The package maps iCalendar dates and
date-times to R's native 'Date' and 'POSIXct' classes.
Recurring items can be expanded.  Timezones are
supported.

* Installation

To install the package from a running R session, type:
#+BEGIN_SRC R :eval never
  install.packages('icalutils, type = 'source',
                   repos = c('http://enricoschumann.net/R',
                             getOption('repos')))
#+END_SRC

or clone/build the GitHub version.


* Examples

#+BEGIN_SRC R :results output :exports both :session **R**
  library("icalutils")
  as.data.frame(read_icalendar("/path/to/file"))
#+END_SRC

Computing recurring events (such as birthdays):

#+BEGIN_SRC R :results output :exports both :session **R**
  as.data.frame(read_calendar("/path/to/file"), 
                recur.expand = TRUE,
                recur.until = Sys.Date + 30) ## events in next 30 days
#+END_SRC

Recurrence sets can be computed with function =rrule=.

#+BEGIN_SRC R :results output :exports both :session **R**
  rrule(dtstart = as.Date("2019-12-01"),
	dtend   = as.Date("2019-12-01"),
	freq = "yearly",
	count = 5)
#+END_SRC

#+RESULTS:
#+begin_example
$text
[1] "FREQ=yearly;COUNT=5;INTERVAL=1"

$recurrence_set
     DTSTART      DTEND
1 2019-12-01 2019-12-01
2 2020-12-01 2020-12-01
3 2021-12-01 2021-12-01
4 2022-12-01 2022-12-01
5 2023-12-01 2023-12-01
#+end_example


Setting up a daily email.

#+begin_src R
  library("icalutils")
  library("mailtools")  ## https://github.com/enricoschumann/mailtools 

  file <- "/path/to/file"
  lookahead <- 30

  ic <- read_icalendar(file, components = "VEVENT")

  df <- as.data.frame(ic, 
		      recur.expand = TRUE,
		      recur.until = Sys.Date() + lookahead)
  df <- df[df$start >= as.POSIXct(paste(Sys.Date(), "00:00:00")) &
	   df$start <  as.POSIXct(Sys.Date() + lookahead), ]

  df <- df[order(df$start), ]
  start <- as.Date(df$start, tz = "")
  end <- as.Date(df$end, tz = "")

  dt <- as.character(start)
  ii <- start != end
  dt[ii] <- paste(as.character(start[ii]), "--", as.character(end[ii]))
  dt

  sendmail(subject = "Kalender",
	   body = c("",
		    paste("Calendar entries for the next", lookahead,  "days"),
		    "",
		    "",
		    paste0(substr(format(df$summary, width = 25), 1, 40),
			   "  [", dt, "]"),
		    ""),

	   method   = "sendemail",
	   to       = "<someone@somewhere.net>",
	   from     = "<someone@somewhere.net>",
	   port     = 587,
	   server   = "<server.net>",
	   user     = "<username>",
	   password = "",
	   headers  = "X-MAIL-CATEGORY: daily calendar report")

#+end_src


NEWS
v0.2-0  (2023-??-??; not yet released)

  o there are now subsetting and print methods for
    "icalendar" and "vevent" objects

  o more rrules are supported

  o additional source repository at
    https://git.sr.ht/~enricoschumann/icalutils

  o 'vevent' now recyles its arguments, allowing
    invokations such as

      vevent(dtstart = c(as.Date("2023-01-28"),
                         as.Date("2023-02-04")),
             summary = "Repeated event")

  o ....

v0.1-0  (2019-10-31)

  o first release: the package provides functions for
    reading files in icalendar format. The package is
    still instable and incomplete; this is an early
    release, mainly for testing purposes.


tests/tinytest.R
if (requireNamespace("tinytest", quietly = TRUE)) {
    tinytest::test_package("icalutils")
}


inst/tinytest/test_RRULE.R
## DAILY
txt <- "FREQ=DAILY;COUNT=10"
DTSTART <- as.Date("2020-10-26")
ans <- RRULE(DTSTART, source = txt)
expect_equal(ans$recurrence_set$DTSTART,
             seq(DTSTART, by = "1 day", length.out = 10))

txt <- "FREQ=DAILY;COUNT=10;INTERVAL=2"
DTSTART <- as.Date("2020-10-26")
ans <- RRULE(DTSTART, source = txt)
expect_equal(ans$recurrence_set$DTSTART,
             seq(DTSTART, by = "2 days", length.out = 10))


## MONTHLY
txt <- "FREQ=MONTHLY;BYDAY=MO,TU,WE,TH,FR;BYSETPOS=-1;UNTIL=20211210"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=MONTHLY;BYDAY=MO,TU,WE,TH,FR;BYSETPOS=1;UNTIL=20211210"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)


txt <- "FREQ=YEARLY;UNTIL=20251210"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=YEARLY;BYMONTH=10;UNTIL=20251210"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=YEARLY;BYMONTH=10,11,12;UNTIL=20251210"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=YEARLY;UNTIL=20251210;INTERVAL=2"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=YEARLY;UNTIL=20251210;INTERVAL=10"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=YEARLY;BYDAY=MO,TU,WE,TH,FR;BYSETPOS=1;UNTIL=20251210"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=YEARLY;BYDAY=MO,TU,WE,TH,FR;BYSETPOS=-1;UNTIL=20251210"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)


## WEEKLY
txt <- "FREQ=WEEKLY;COUNT=10"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=WEEKLY;COUNT=10;INTERVAL=2"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=WEEKLY;COUNT=10;INTERVAL=2;BYDAY=MO,TU"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)

txt <- "FREQ=WEEKLY;COUNT=10;INTERVAL=2;BYDAY=MO,TU;WKST=TU"
DTSTART <- as.Date("2020-10-26")
RRULE(DTSTART, source = txt)



## YEARLY
txt <- "FREQ=YEARLY;INTERVAL=1"
DTSTART <- as.Date("1978-08-19")
RRULE(DTSTART, source = txt)

txt <- "FREQ=YEARLY;INTERVAL=2;COUNT=10"
DTSTART <- as.Date("1978-08-19")
RRULE(DTSTART, source = txt)

txt <- "FREQ=YEARLY;COUNT=10"
DTSTART <- as.Date("1978-08-19")
RRULE(DTSTART, source = txt)

txt <- "FREQ=YEARLY"
DTSTART <- as.Date("1978-08-19")
RRULE(DTSTART, source = txt)


R/tznames.R
## The data in this file are auto-generated from file 'windowsZones.xml'
## in the Unicode Common Locale Data Repository (http://cldr.unicode.org/).
## See https://www.unicode.org/copyright.html and https://www.unicode.org/license.html.

.tznames <- structure(list(Windows = c("Dateline Standard Time",
    "UTC-11", "Aleutian Standard Time",
    "Hawaiian Standard Time",
    "Marquesas Standard Time",
    "Alaskan Standard Time",
    "UTC-09", "Pacific Standard Time (Mexico)",
    "UTC-08", "Pacific Standard Time",
    "US Mountain Standard Time",
    "Mountain Standard Time (Mexico)",
    "Mountain Standard Time",
    "Yukon Standard Time",
    "Central America Standard Time",
    "Central Standard Time",
    "Easter Island Standard Time",
    "Central Standard Time (Mexico)",
    "Canada Central Standard Time",
    "SA Pacific Standard Time",
    "Eastern Standard Time (Mexico)",
    "Eastern Standard Time",
    "Haiti Standard Time",
    "Cuba Standard Time",
    "US Eastern Standard Time",
    "Turks And Caicos Standard Time",
    "Paraguay Standard Time",
    "Atlantic Standard Time",
    "Venezuela Standard Time",
    "Central Brazilian Standard Time",
    "SA Western Standard Time",
    "Pacific SA Standard Time",
    "Newfoundland Standard Time",
    "Tocantins Standard Time",
    "E. South America Standard Time",
    "SA Eastern Standard Time",
    "Argentina Standard Time",
    "Greenland Standard Time",
    "Montevideo Standard Time",
    "Magallanes Standard Time",
    "Saint Pierre Standard Time",
    "Bahia Standard Time",
    "UTC-02", "Azores Standard Time",
    "Cape Verde Standard Time",
    "UTC", "GMT Standard Time",
    "Greenwich Standard Time",
    "Sao Tome Standard Time",
    "Morocco Standard Time",
    "W. Europe Standard Time",
    "Central Europe Standard Time",
    "Romance Standard Time",
    "Central European Standard Time",
    "W. Central Africa Standard Time",
    "Jordan Standard Time",
    "GTB Standard Time",
    "Middle East Standard Time",
    "Egypt Standard Time",
    "E. Europe Standard Time",
    "Syria Standard Time",
    "West Bank Standard Time",
    "South Africa Standard Time",
    "FLE Standard Time",
    "Israel Standard Time",
    "South Sudan Standard Time",
    "Kaliningrad Standard Time",
    "Sudan Standard Time",
    "Libya Standard Time",
    "Namibia Standard Time",
    "Arabic Standard Time",
    "Turkey Standard Time",
    "Arab Standard Time",
    "Belarus Standard Time",
    "Russian Standard Time",
    "E. Africa Standard Time",
    "Iran Standard Time",
    "Arabian Standard Time",
    "Astrakhan Standard Time",
    "Azerbaijan Standard Time",
    "Russia Time Zone 3",
    "Mauritius Standard Time",
    "Saratov Standard Time",
    "Georgian Standard Time",
    "Volgograd Standard Time",
    "Caucasus Standard Time",
    "Afghanistan Standard Time",
    "West Asia Standard Time",
    "Ekaterinburg Standard Time",
    "Pakistan Standard Time",
    "Qyzylorda Standard Time",
    "India Standard Time",
    "Sri Lanka Standard Time",
    "Nepal Standard Time",
    "Central Asia Standard Time",
    "Bangladesh Standard Time",
    "Omsk Standard Time",
    "Myanmar Standard Time",
    "SE Asia Standard Time",
    "Altai Standard Time",
    "W. Mongolia Standard Time",
    "North Asia Standard Time",
    "N. Central Asia Standard Time",
    "Tomsk Standard Time",
    "China Standard Time",
    "North Asia East Standard Time",
    "Singapore Standard Time",
    "W. Australia Standard Time",
    "Taipei Standard Time",
    "Ulaanbaatar Standard Time",
    "Aus Central W. Standard Time",
    "Transbaikal Standard Time",
    "Tokyo Standard Time",
    "North Korea Standard Time",
    "Korea Standard Time",
    "Yakutsk Standard Time",
    "Cen. Australia Standard Time",
    "AUS Central Standard Time",
    "E. Australia Standard Time",
    "AUS Eastern Standard Time",
    "West Pacific Standard Time",
    "Tasmania Standard Time",
    "Vladivostok Standard Time",
    "Lord Howe Standard Time",
    "Bougainville Standard Time",
    "Russia Time Zone 10",
    "Magadan Standard Time",
    "Norfolk Standard Time",
    "Sakhalin Standard Time",
    "Central Pacific Standard Time",
    "Russia Time Zone 11",
    "New Zealand Standard Time",
    "UTC+12", "Fiji Standard Time",
    "Chatham Islands Standard Time",
    "UTC+13", "Tonga Standard Time",
    "Samoa Standard Time",
    "Line Islands Standard Time"),
    Olson = c("Etc/GMT+12",
        "Etc/GMT+11",
        "America/Adak",
        "Pacific/Honolulu",
        "Pacific/Marquesas",
        "America/Anchorage",
        "Etc/GMT+9",
        "America/Tijuana",
        "Etc/GMT+8",
        "America/Los_Angeles",
        "America/Phoenix",
        "America/Chihuahua",
        "America/Denver",
        "America/Whitehorse",
        "America/Guatemala",
        "America/Chicago",
        "Pacific/Easter",
        "America/Mexico_City",
        "America/Regina",
        "America/Bogota",
        "America/Cancun",
        "America/New_York",
        "America/Port-au-Prince",
        "America/Havana",
        "America/Indianapolis",
        "America/Grand_Turk",
        "America/Asuncion",
        "America/Halifax",
        "America/Caracas",
        "America/Cuiaba",
        "America/La_Paz",
        "America/Santiago",
        "America/St_Johns",
        "America/Araguaina",
        "America/Sao_Paulo",
        "America/Cayenne",
        "America/Buenos_Aires",
        "America/Godthab",
        "America/Montevideo",
        "America/Punta_Arenas",
        "America/Miquelon",
        "America/Bahia",
        "Etc/GMT+2",
        "Atlantic/Azores",
        "Atlantic/Cape_Verde",
        "Etc/UTC", "Europe/London",
        "Atlantic/Reykjavik",
        "Africa/Sao_Tome",
        "Africa/Casablanca",
        "Europe/Berlin",
        "Europe/Budapest",
        "Europe/Paris",
        "Europe/Warsaw",
        "Africa/Lagos",
        "Asia/Amman",
        "Europe/Bucharest",
        "Asia/Beirut",
        "Africa/Cairo",
        "Europe/Chisinau",
        "Asia/Damascus",
        "Asia/Hebron",
        "Africa/Johannesburg",
        "Europe/Kiev",
        "Asia/Jerusalem",
        "Africa/Juba",
        "Europe/Kaliningrad",
        "Africa/Khartoum",
        "Africa/Tripoli",
        "Africa/Windhoek",
        "Asia/Baghdad",
        "Europe/Istanbul",
        "Asia/Riyadh",
        "Europe/Minsk",
        "Europe/Moscow",
        "Africa/Nairobi",
        "Asia/Tehran",
        "Asia/Dubai",
        "Europe/Astrakhan",
        "Asia/Baku",
        "Europe/Samara",
        "Indian/Mauritius",
        "Europe/Saratov",
        "Asia/Tbilisi",
        "Europe/Volgograd",
        "Asia/Yerevan",
        "Asia/Kabul",
        "Asia/Tashkent",
        "Asia/Yekaterinburg",
        "Asia/Karachi",
        "Asia/Qyzylorda",
        "Asia/Calcutta",
        "Asia/Colombo",
        "Asia/Katmandu",
        "Asia/Almaty",
        "Asia/Dhaka",
        "Asia/Omsk",
        "Asia/Rangoon",
        "Asia/Bangkok",
        "Asia/Barnaul",
        "Asia/Hovd",
        "Asia/Krasnoyarsk",
        "Asia/Novosibirsk",
        "Asia/Tomsk",
        "Asia/Shanghai",
        "Asia/Irkutsk",
        "Asia/Singapore",
        "Australia/Perth",
        "Asia/Taipei",
        "Asia/Ulaanbaatar",
        "Australia/Eucla",
        "Asia/Chita",
        "Asia/Tokyo",
        "Asia/Pyongyang",
        "Asia/Seoul",
        "Asia/Yakutsk",
        "Australia/Adelaide",
        "Australia/Darwin",
        "Australia/Brisbane",
        "Australia/Sydney",
        "Pacific/Port_Moresby",
        "Australia/Hobart",
        "Asia/Vladivostok",
        "Australia/Lord_Howe",
        "Pacific/Bougainville",
        "Asia/Srednekolymsk",
        "Asia/Magadan",
        "Pacific/Norfolk",
        "Asia/Sakhalin",
        "Pacific/Guadalcanal",
        "Asia/Kamchatka",
        "Pacific/Auckland",
        "Etc/GMT-12",
        "Pacific/Fiji",
        "Pacific/Chatham",
        "Etc/GMT-13",
        "Pacific/Tongatapu",
        "Pacific/Apia",
        "Pacific/Kiritimati")),
    class = "data.frame",
    row.names = c(NA,
        -139L))


R/utilities.R
## copied from datetimeutils https://github.com/enricoschumann/datetimeutils

.mday <- function(x)
    as.POSIXlt(x)$mday

.year <- function(x)
    as.POSIXlt(x)$year + 1900

.month <- function(x)
    as.POSIXlt(x)$mon + 1

.week <- function(x, WKST = 1) {

    ## RFC 5545, 3.3.10.
    ## 
    ## Week number one of the calendar year is the first
    ## week that contains at least four (4) days in that
    ## calendar year.

    if (!inherits(x, "Date"))
        x <- as.Date(as.POSIXlt(x))

    x <- (x + 4 - ((.weekday(x) - WKST) %% 7 +1))
    y1 <- x - as.POSIXlt(x)$yday
    as.numeric(x - y1) %/% 7+1
}

.start_of_week <- function(x, WKST = 1, Date = TRUE) {

    ## map a given date to the start of its week
    offset <- if (WKST == 0)
                  3
              else
                  WKST - 4

    if (Date)
        .Date(7 * (unclass(x - offset) %/% 7) + offset)
    else
              7 * (unclass(x - offset) %/% 7) + offset
}

.weekday <- function(x)
    ## maps Date to a number (0 Sunday .. 6 Saturday)
    unclass(x + 4) %% 7

.wday <- c(
    "SU" = 0,
    "MO" = 1,
    "TU" = 2,
    "WE" = 3,
    "TH" = 4,
    "FR" = 5,
    "SA" = 6)

.weekdayS <- function(dates)
        c("SU",
          "MO",
          "TU",
          "WE",
          "TH",
          "FR",
          "SA")[unclass(dates + 4) %% 7 + 1]

.next_weekday <- function(wday, start, count = 1, interval = 1) {
    ## wday : 0 SU, 1 MO ... => target day
    ## start: a date
    if (count == 0)
        return(numeric(0L))
    ans <- (wday - .weekday(start)) %% 7 +
        rep(7 * interval * seq(0, count - 1), each = length(wday))
    ans <- sort(ans) + start
    ans[seq_len(count)]
}



R/ical_utils.R
.eol.strict     <- "\r\n"
.fold.re        <- "\n[ \t]"
.fold.re.strict <- "\r\n[ \t]"

.fold <- function(s) {}

.unfold <- function(s, strict.eol) {}

.read_one_line <- function(filename, nchars = 200) {
    zz <- file(filename, "rb")
    on.exit(close(zz))
    vcal <- list()
    i <- 0
    while(length(txt <- readChar(zz, nchars))) {
        i <- i + 1
        vcal[[i]] <- txt
    }
    paste(unlist(vcal), collapse = "")
}

read_icalendar <- function(file, ...) {
    c(read_vevent(file),
      read_vtimezone(file))
}

read_vevent <- function(file, ...,
                        strict.eol = TRUE,
                        adjust.allday = TRUE,
                        return.class = "list",
                        recur.expand = FALSE,
                        recur.until = NULL,
                        recur.count = NULL,
                        timestamps.POSIXct = TRUE,
                        timestamps.Date = FALSE,
                        timestamps.tz = "") {


    cal <- .read_one_line(file)
    cal <- gsub(if (strict.eol) .fold.re.strict else .fold.re, "", cal)
    cal <- strsplit(cal, "\r?\n")[[1]]


    ## rfc5545: VEVENTs cannot be nested
    begin <- grep("^BEGIN:VEVENT", cal)
    end <- grep("^END:VEVENT", cal)

    res <- list()
    for (i in seq_along(begin)) {
        event <- cal[seq(begin[i] + 1, end[i] - 1)]
        res[[i]] <- .expand_properties(.properties(event))
    }
    if (return.class == "data.frame") {
        uid <- unlist(lapply(res, `[[`, "UID"))
        recurring <- unlist(lapply(res, function(x) "RRULE" %in% names(x)))

        start <- lapply(res, `[[`, "DTSTART")
        end <- lapply(res, `[[`, "DTEND")
        if (recur.expand) {
            RRULE <- lapply(res[recurring], `[[`, "RRULE")
            for (r in seq_along(RRULE)) {
                created <- icalutils:::.seq_rrule(start[recurring][[r]],
                                                  end  [recurring][[r]],
                                                  RRULE[[r]],
                                                  UNTIL = recur.until,
                                                  COUNT = recur.count)
                if (isFALSE(created))
                    message("cannot parse RRULE",
                            RRULE[[r]])

            }
        }

        all.day <- unlist(lapply(start, function(x) inherits(x, "Date")))
        start[all.day] <- lapply(start[all.day],
                                 function(x) as.POSIXct(paste(x, "00:00:00"),
                                                        tz = timestamps.tz))
        start <- .POSIXct(unlist(start))

        all.day <- unlist(lapply(end, function(x) inherits(x, "Date")))
        end[all.day] <- lapply(end[all.day],
                               function(x) as.POSIXct(paste(x, "00:00:00"),
                                                      tz = timestamps.tz))
        end <- .POSIXct(unlist(end))

        summary <- lapply(res, `[[`, "SUMMARY")
        summary <- unlist(lapply(summary,
                              function(x) if (is.null(x)) "" else x))

        description <- lapply(res, `[[`, "DESCRIPTION")
        description <- unlist(lapply(description,
                              function(x) if (is.null(x)) "" else x))

        res <- data.frame(uid,
                          summary,
                          description,
                          start,
                          end,
                          all.day,
                          recurring,
                          stringsAsFactors = FALSE)
    }
    res
}

read_vtimezone <- function(file, ...,
                        strict.eol = TRUE) {
    cal <- .read_one_line(file)
    cal <- gsub(if (strict.eol) .fold.re.strict else .fold.re, "", cal)
    cal <- strsplit(cal, "\r?\n")[[1]]

    ## rfc5545: VTIMEZONE cannot be nested
    begin <- grep("^BEGIN:VTIMEZONE", cal)
    end <- grep("^END:VTIMEZONE", cal)

    res <- list()
    for (i in seq_along(begin)) {
        event <- cal[seq(begin[i]+1, end[i]-1)]
        res[[i]] <- .expand_properties(.properties(event))
    }
    res
}

seq_rrule <- function(start, event, until = NULL, count = NULL) {

}

.seq_rrule <- function(DTSTART, DTEND, RRULE, UNTIL, COUNT,
                       msg.violations = TRUE) {
    ## FREQ, ## MUST
    ## UNTIL,
    ## COUNT,
    ## INTERVAL = 1,
    ## BYSECOND,
    ## BYMINUTE,
    ## BYHOUR,
    ## BYDAY,
    ## BYMONTHDAY,  ## 1 -- 31    -1 -- -31
    ## BYYEARDAY,   ## 1 -- 366   -1 -- 366
    ## BYWEEKNO,    ## 1 -- 53
    ## BYMONTH,  ## 1 - 12
    ## BYSETPOS,
    ## WKST   ## weekstart: MO, TU, WE, TH, FR, SA, and SU

    ## browser()
    FREQ <- toupper(RRULE$FREQ)

    if (is.null(RRULE$INTERVAL))
        INTERVAL <- 1
    else
        INTERVAL <- RRULE$INTERVAL

    if (inherits(DTSTART, "POSIXct"))
        DTSTARTlt <- as.POSIXlt(DSTART)

    if (!is.null(RRULE$UNTIL)) {
        UNTIL <- RRULE$UNTIL
        if (!class(DTSTART) %in% class(UNTIL) && msg.violations) {
            message("DTSTART and UNTIL are not of the same type [3.3.10.]")
            if (inherits(DTSTART, "Date"))
                UNTIL <- as.Date(UNTIL)
        }
    } else if (!is.null(RRULE$COUNT)) {
        COUNT <- RRULE$COUNT
    } else if (is.null(RRULE$UNTIL) &&
               is.null(RRULE$COUNT) &&
               is.null(UNTIL) &&
               is.null(COUNT)) {
        ## neither COUNT nor UNTIL are specified in RRULE,
        ## and recur.count/recur.until are NULL

        if (inherits(DTSTART, "Date"))
            UNTIL <- Sys.Date() + 365*10
        else {
            UNTIL <- as.POSIXct(Sys.Date() + 365*10)
        }
        message("no UNTIL specified: use ", UNTIL)
    }

    ## "the last work day of the month"
    ## FREQ=MONTHLY;BYDAY=MO,TU,WE,TH,FR;BYSETPOS=-1

    ## "last day of a month"
    ## FREQ=MONTHLY;BYDAY=MO,TU,WE,TH,FR,SA,SU;BYSETPOS=-1

    ## "SECONDLY" / "MINUTELY" / "HOURLY" / "DAILY"
    ## / "WEEKLY" / "MONTHLY" / "YEARLY"
    ## => each can be of length > 1

    ## BYDAY "SU" / "MO" / "TU" / "WE" / "TH" / "FR" / "SA"
    ## +1SU first sunday, -1SU last sunday ## MONTHLY and YEARLY

    .wday <- c(
        "SU" = 0,
        "MO" = 1,
        "TU" = 2,
        "WE" = 3,
        "TH" = 4,
        "FR" = 5,
        "SA" = 6)

    ans <- FALSE
    if (FREQ == "YEARLY") {
        if        ( is.null(RRULE$BYSECOND)   &&
                    is.null(RRULE$BYMINUTE)   &&
                    is.null(RRULE$BYHOUR)     &&
                    is.null(RRULE$BYDAY)      &&
                    is.null(RRULE$BYMONTHDAY) &&
                    is.null(RRULE$BYYEARDAY)  &&
                    is.null(RRULE$BYWEEKNO)   &&
                    is.null(RRULE$BYMONTH)    &&
                    is.null(RRULE$BYSETPOS)   &&
                    is.null(RRULE$WKST)) {

            if (inherits(DTSTART, "Date"))
                ans <- seq(DTSTART, to = UNTIL, by = paste(INTERVAL, "year"))


        } else if ( is.null(RRULE$BYSECOND)   &&
                    is.null(RRULE$BYMINUTE)   &&
                    is.null(RRULE$BYHOUR)     &&
                   !is.null(RRULE$BYDAY)      &&
                    is.null(RRULE$BYMONTHDAY) &&
                    is.null(RRULE$BYYEARDAY)  &&
                    is.null(RRULE$BYWEEKNO)   &&
                   !is.null(RRULE$BYMONTH)    &&
                    is.null(RRULE$BYSETPOS)   &&
                    is.null(RRULE$WKST)) {



        }
    } else if (FREQ == "MONTHLY") {
        if        ( is.null(RRULE$BYSECOND)   &&
                    is.null(RRULE$BYMINUTE)   &&
                    is.null(RRULE$BYHOUR)     &&
                    is.null(RRULE$BYDAY)      &&
                    is.null(RRULE$BYMONTHDAY) &&
                    is.null(RRULE$BYYEARDAY)  &&
                    is.null(RRULE$BYWEEKNO)   &&
                   !is.null(RRULE$BYMONTH)    &&
                    is.null(RRULE$BYSETPOS)   &&
                    is.null(RRULE$WKST)) {



        }
    } else if (FREQ == "WEEKLY") {
        if        ( is.null(RRULE$BYSECOND)   &&
                    is.null(RRULE$BYMINUTE)   &&
                    is.null(RRULE$BYHOUR)     &&
                   !is.null(RRULE$BYDAY)      &&
                    is.null(RRULE$BYMONTHDAY) &&
                    is.null(RRULE$BYYEARDAY)  &&
                    is.null(RRULE$BYWEEKNO)   &&
                    is.null(RRULE$BYMONTH)    &&
                    is.null(RRULE$BYSETPOS)   &&
                    is.null(RRULE$WKST)) {

            if (inherits(DTSTART, "Date")) {
                if (is.null(COUNT))
                    COUNT <- ceiling(as.numeric(UNTIL - DTSTART)/(7*INTERVAL))

                ans <- .next_weekday(.wday[RRULE$BYDAY],
                                     start = DTSTART,
                                     count = COUNT,
                                     interval = INTERVAL)
                ans <- ans[ans <= UNTIL]

            } else {
                lt <- as.POSIXlt(DTSTART)
                dates <- .next_weekday(.wday[RRULE$BYDAY],
                                       start = as.Date(DTSTART),
                                       count = COUNT,
                                       interval = INTERVAL)
                ans <- as.POSIXct(paste(dates, lt$hour, lt$min, lt$sec),
                                  format = "%Y-%m-%d %H %M %S",
                                  tzone = attributes(DTSTART, "tzone"))
            }
        }
    }

    ## TODO: remove EXDATEs

    ans
}


.next_weekday <- function(wday, start, count = 1, interval = 1)
    start + wday - unclass(start + 4) %% 7 +
        interval*7L*(seq_len(count) - 1L)


.utc_dt <- function(x, ...) {
    ## trailing Z does not have to be removed
    as.POSIXct(strptime(x,
                        format = "%Y%m%dT%H%M%S",
                        tz = "UTC"))
}

.parse_rrule <- function(rrule, ...) {

    ## takes a vector of one or more rrules and
    ## returns a list
    rules <- strsplit(rrule, ";", fixed = TRUE)
    rules <- lapply(rules, strsplit, "=", fixed = TRUE)
    r.names <- lapply(rules, function(x) lapply(x, `[[`, 1))
    r.names <- lapply(r.names, unlist)
    rules <- lapply(rules, function(x) lapply(x, `[`, -1))

    rules <- mapply(`names<-`, rules, r.names, SIMPLIFY = FALSE)

    rules <- lapply(rules,
                    function(x) {
        nm <- names(x)
        if ("INTERVAL" %in% nm)
            x[["INTERVAL"]] <- as.numeric(x[["INTERVAL"]])
        else
            x[["INTERVAL"]] <- 1

        if ("UNTIL" %in% nm)  ## FIXME: if DATE?
            x[["UNTIL"]] <- .utc_dt(x[["UNTIL"]])
        class(x) <- "ical_utils_rrule"
        x
    })

    rules
}

.properties <- function(s, ...) {
    ## receives a character vector (lines of iCalendar),
    ## and returns named character vector (names = properties)
    p <- "^([^:;]+?)[:;](.*)"
    ans <- gsub(p, "\\2", s, perl = TRUE)
    names(ans) <- gsub(p, "\\1", s, perl = TRUE)
    ans
}

.expand_properties <- function(p) {
    ## receives a named character vector (names = properties),
    ## and returns a list
    ans <- as.list(p)
    nm <- names(ans)
    dtfields <- c("CREATED",
                  "LAST-MODIFIED",
                  "DTSTAMP",
                  "DTSTART",
                  "DTEND",
                  "DUE")
    for (field in dtfields) {
        if (field %in% nm) {
            if (endsWith(ans[[field]], "Z")) {
                ans[[field]] <- as.POSIXct(
                    strptime(ans[[field]],
                             format = "%Y%m%dT%H%M%S",
                             tz = "UTC"))
            } else if (grepl("^VALUE=DATE",
                             ans[[field]],
                             ignore.case = TRUE)) {
                n <- nchar(ans[[field]])
                ans[[field]] <- as.Date(substr(ans[[field]], n - 7, n),
                                        format = "%Y%m%d")
            } else if (grepl("^TZID",
                             ans[[field]],
                             ignore.case = TRUE)) {
                n <- nchar(ans[[field]])
                tz <- substr(ans[[field]], 6, n - 16)
                ans[[field]] <- as.POSIXct(
                    strptime(substr(ans[[field]], n - 14, n),
                             format = "%Y%m%dT%H%M%S",
                             tz = tz))
            }
        }
    }
    if ("RRULE" %in% nm) {
        ans[["RRULE"]] <- .parse_rrule(ans[["RRULE"]])[[1]]

    }
    ans
}


ical_structure <- function(file, ..., strict.eol = TRUE) {
    cal <- .read_one_line(file)
    cal <- gsub(if (strict.eol) .fold.re.strict else .fold.re, "", cal)
    cal <- strsplit(cal, "\r?\n")[[1]]

    n <- length(cal)
    component <- character(n)
    level <-start <- end <- numeric(n)
    current.level <- 0
    for (i in seq_along(cal)) {
        if (grepl("^BEGIN:", cal[i])) {
            current.level <- current.level + 1
            cat(rep("  ", current.level-1), cal[i], "\n", sep = "")
            level[i] <- current.level
            start[i] <- i
            component[i] <- cal[i]
        }
        if (grepl("^END:", cal[i])) {
            current.level <- current.level - 1
            end[i] <- i
        }
    }
    data.frame(component, level, start, end)
}

allday_event <- function(date, summary, description = "", file) {

    UID <- paste0(.msgID(), ".", Sys.info()["nodename"])
    DTSTAMP <- format(as.POSIXlt(Sys.time(), tz = "UTC"), "%Y%m%dT%H%M%SZ")

    DTSTART <- format(date, "%Y%m%d")

    version <- packageVersion("ical.utils")
tmp <-
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//enricoschumann.net/R/packages/ical.utils//NONSGML ical.utils {version}//EN
BEGIN:VEVENT
UID:{UID}
DTSTAMP:{DTSTAMP}
DTSTART;VALUE=DATE:{DTSTART}
SUMMARY:{SUMMARY}
DESCRIPTION:{DESCRIPTION}
END:VEVENT
END:VCALENDAR"

    tmp <- fill_in(tmp,
                   UID = UID,
                   DTSTAMP = DTSTAMP,
                   DTSTART = DTSTART,
                   SUMMARY = summary,
                   DESCRIPTION = description,
                   version = version)
    tmp <- strsplit(tmp, "\n", fixed = TRUE)[[1L]]
    tmp <- paste(tmp, collapse = .eol.strict)
    if (!missing(file)) {
        writeLines(tmp, file)
        invisible(tmp)
    } else
        tmp

}

## copy of https://github.com/enricoschumann/mailtools/blob/master/R/msgID.R
.msgID <- function() {
    digits <- function(k, b) {
        if (all(k == 0L))
            return(rep(0, length(k)))
        nd <- 1 + floor(log(max(k), b))
        ans <- numeric(length(k) * nd)
        dim(ans) <- c(length(k), nd)
        for (i in nd:1) {
            ans[ ,i] <- k %% b
            if (i > 1L)
                k <- k%/%b
        }
        ans
    }
    ab <- c(1:9, 0,letters[1:26])
    n <- c(floor(runif(1)*1e16),
           floor(10000*as.numeric(Sys.time())))
    paste(c(ab[digits(n[1L], 36) + 1], ".",
            ab[digits(n[2L], 36) + 1]), sep = "", collapse="")
}
